RecoLocalCalo/HcalRecAlgos/interface/MahiFit.h

```diff
@@ -0,0 +1,165 @@
+#ifndef RecoLocalCalo_HcalRecAlgos_MahiFit_HH
+#define RecoLocalCalo_HcalRecAlgos_MahiFit_HH
+
+#include "DataFormats/HcalDetId/interface/HcalSubdetector.h"
+#include "DataFormats/HcalDetId/interface/HcalDetId.h"
+#include "CalibFormats/HcalObjects/interface/HcalCalibrations.h"
+#include "CalibFormats/HcalObjects/interface/HcalCoder.h"
+#include "RecoLocalCalo/HcalRecAlgos/interface/EigenMatrixTypes.h"
+#include "DataFormats/HcalRecHit/interface/HBHEChannelInfo.h"
+
+#include "CalibCalorimetry/HcalAlgos/interface/HcalPulseShapes.h"
+#include "CalibCalorimetry/HcalAlgos/interface/HcalTimeSlew.h"
+#include "RecoLocalCalo/HcalRecAlgos/interface/PulseShapeFunctor.h"
+
+#include <Math/Functor.h>
+
+struct nnlsWorkspace {
+
+  unsigned int nPulseTot;
+  unsigned int tsSize;
+  unsigned int tsOffset;
+  unsigned int fullTSOffset;
+  int bxOffset;
+  double dt;
+
+  //holds active bunch crossings
+  BXVector bxs;  
+
+  //holds data samples
+  SampleVector amplitudes;
+
+  //holds inverse covariance matrix
+  SampleMatrix invCovMat;
+
+  //holds diagonal noise terms
+  SampleVector noiseTerms;
+
+  //holds flat pedestal uncertainty
+  SampleMatrix pedConstraint;
+  
+  //holds full covariance matrix for a pulse shape 
+  //varied in time
+  std::array<FullSampleMatrix, MaxPVSize> pulseCovArray;
+
+  //holds full pulse shape template
+  std::array<FullSampleVector, MaxPVSize> pulseShapeArray;
+
+  //holds full pulse shape derivatives
+  std::array<FullSampleVector, MaxPVSize> pulseDerivArray;
+
+  //holders for calculating pulse shape & covariance matrices
+  std::array<double, MaxSVSize> pulseN;
+  std::array<double, MaxSVSize> pulseM;
+  std::array<double, MaxSVSize> pulseP;
+
+  //holds matrix of pulse shape templates for each BX
+  SamplePulseMatrix pulseMat;
+
+  //holds matrix of pulse shape derivatives for each BX
+  SamplePulseMatrix pulseDerivMat;
+
+  //holds residual vector
+  PulseVector residuals;
+
+  //for FNNLS algorithm
+  unsigned int nP;
+  PulseVector ampVec;
+
+  PulseVector errVec;
+  PulseVector ampvecpermtest;
+
+  SamplePulseMatrix invcovp;
+  PulseMatrix aTaMat; // A-transpose A (matrix)
+  PulseVector aTbVec; // A-transpose b (vector)
+  PulseVector updateWork; // w (vector)
+
+  SampleDecompLLT covDecomp;
+  SampleMatrix covDecompLinv;
+  PulseMatrix topleft_work;
+  PulseDecompLDLT pulseDecomp;
+
+};
```

Comment:
to follow up on  https://github.com/cms-sw/cmssw/pull/21664#discussion_r156724821

> Perhaps an "almost const" solution would be to have data members refactored, e.g. make a state struct (even on heap) with a clear reset call at the end/entry of phase1Apply.

I do not see a reset method in the current implementation.

> All other methods called by phase1Apply will be const and operate on this state struct.

This implied that the struct is passed by argument

Reply:
Ah,I may have misinterpreted this request.This is what i did to reset everything:  https://github.com/jaylawhorn/cmssw/blob/314d16ffe353ff1e6719567a146dde12c92747b9/RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc#L600

Reply:
ah, it's kind of covered then, I should have read the implementation in .cc